ctru-rs/examples/camera-image.rs

```diff
@@ -0,0 +1,201 @@
+};
+use ctru::services::hid::KeyPad;
+use ctru::services::{Apt, Hid};
+use ctru::Gfx;
+
+const WIDTH: usize = 400;
+const HEIGHT: usize = 240;
+const SCREEN_SIZE: usize = WIDTH * HEIGHT * 2;
+const BUF_SIZE: usize = SCREEN_SIZE * 2;
+
+const WAIT_TIMEOUT: i64 = 300000000;
+
+fn main() {
+    ctru::init();
+
+    let apt = Apt::init().expect("Failed to initialize Apt service.");
+    let hid = Hid::init().expect("Failed to initialize Hid service.");
+    let gfx = Gfx::init().expect("Failed to initialize GFX service.");
+
+    gfx.top_screen.borrow_mut().set_double_buffering(true);
+    gfx.bottom_screen.borrow_mut().set_double_buffering(false);
+
+    let _console = Console::init(gfx.bottom_screen.borrow_mut());
+
+    let mut key_down;
+    let mut key_held;
+
+    println!("Initializing camera");
+
+    let mut cam = Cam::init().expect("Failed to initialize CAM service.");
+
+    cam.set_size(
+        CamSelect::SELECT_OUT1_OUT2,
+        CamSize::SIZE_CTR_TOP_LCD,
+        CamContext::CONTEXT_A,
+    )
+    .expect("Failed to set camera size");
+    cam.set_output_format(
+        CamSelect::SELECT_OUT1_OUT2,
+        CamOutputFormat::OUTPUT_RGB_565,
+        CamContext::CONTEXT_A,
+    )
+    .expect("Failed to set camera output format");
+
+    cam.set_noise_filter(CamSelect::SELECT_OUT1_OUT2, true)
+        .expect("Failed to enable noise filter");
+    cam.set_auto_exposure(CamSelect::SELECT_OUT1_OUT2, true)
+        .expect("Failed to enable auto exposure");
+    cam.set_auto_white_balance(CamSelect::SELECT_OUT1_OUT2, true)
+        .expect("Failed to enable auto white balance");
+
+    cam.set_trimming(CamPort::PORT_CAM1, false)
+        .expect("Failed to disable trimming for Cam Port 1");
+    cam.set_trimming(CamPort::PORT_CAM2, false)
+        .expect("Failed to disable trimming for Cam Port 2");
+
+    let mut buf = vec![0u8; BUF_SIZE];
+
+    gfx.flush_buffers();
+    gfx.swap_buffers();
+    gfx.wait_for_vblank();
+
+    let mut held_r = false;
+
+    println!("\nPress R to take a new picture");
+    println!("Press Start to exit to Homebrew Launcher");
+
+    gfx.top_screen.borrow_mut().set_3d_enabled(false);
+
+    while apt.main_loop() {
+        hid.scan_input();
+        key_down = hid.keys_down();
+        key_held = hid.keys_held();
+
+        if key_down.contains(KeyPad::KEY_START) {
+            break;
+        }
+
+        if key_held.contains(KeyPad::KEY_R) && !held_r {
+            println!("Capturing new image");
+            gfx.flush_buffers();
+            gfx.swap_buffers();
+            gfx.wait_for_vblank();
+            held_r = true;
+            take_picture(&mut cam, &mut buf);
+        } else if !key_held.contains(KeyPad::KEY_R) {
+            held_r = false;
+        }
+
+        write_picture_to_frame_buffer_rgb_565(
+            gfx.top_screen.borrow_mut().get_raw_framebuffer(Side::Left),
+            &mut buf,
+            0,
+            0,
+            WIDTH as u16,
```

Comment:
Other than this specific line, there are many instances of `as i16` or `as u32`. Is there nothing to be done about it? (maybe the API is setup with the wrong parameters?)

Reply:
The issue isn't the API, more that numbers with different types are used in different kinds of math. `ctr-sys` takes the types they are casted to which I thought may want to stay consistent. If they are individually defined without reusing each other to determine values the need for casting could be removed.